feat(shared.lua): `vim.traverse`

llllvvuu opened the pull request:
Might be jumping the gun on this (not very critical), but it was low-cost to just post it.

It replaces the lost recursive functionality from #25118 and mainly it would just exist to nudge people away from coming up with something worse.

llllvvuu committed:
feat(shared.lua): `vim.traverse`

Might be jumping the gun on this (not very critical), but it was
low-cost to just post it.

It replaces the lost recursive functionality from #25118 and mainly it
would just exist to nudge people away from coming up with something
worse.

clason commented:
>  although this means the optional arg isn't last.

General API guideline is:
* mandatory arguments are positional
* optional arguments are keyword-style (i.e., through a final `opts` table)

clason commented:
Just going by the documentation, that looks to be rather specific to treesitter -- too much to be under the general `vim` API.

I suspect that the function itself is much more general, and would recommend that the documentation reflect that ("traverse an arbitrary tree defined by ..."), with treesitter tree used as an example.

justinmk commented:
> > although this means the optional arg isn't last.
> 
> General API guideline is:
> 
> * mandatory arguments are positional
> * optional arguments are keyword-style (i.e., through a final `opts` table)

Need to amend that. The callback should always be last. `opts` doesn't to be last. It's much more readable if the callback is the final arg, otherwise one must travel to the end of the callback body to see if additional args were passed.

clason commented:
In this case, we can just make `key` mandatory, though; I don't see much benefit of allowing (explicit) `nil` for a (treesitter specific) default.

lucario387 commented on runtime/doc/lua.txt:
```diff
@@ -2323,6 +2323,18 @@ vim.tbl_values({t})                                         *vim.tbl_values()*
+
+    Parameters: ~
+      • {node}     (table) The root node to visit
+      • {key}      nil|string|fun(node: Node): Node[] How to find the children
```
I don't think a "default" for a key makes sense here

llllvvuu commented:
> Just going by the documentation, that looks to be rather specific to treesitter

Reworded

> we can just make key mandatory

Good, that solves the dilemma. Updated.

I also removed the generic since it doesn't seem to work well with the doc pipeline (or Lua in general)

clason commented:
I think a concrete example would be helpful then, though?

dundargoc commented:
> > > although this means the optional arg isn't last.
> > 
> > 
> > General API guideline is:
> > 
> > * mandatory arguments are positional
> > * optional arguments are keyword-style (i.e., through a final `opts` table)
> 
> Need to amend that. The callback should always be last. `opts` can be penultimate. It's much more readable if the callback is the final arg, otherwise one must travel to the end of the callback body to see if additional args were passed.

@justinmk mind adding that to `:h dev-api`? These things are hard to remember later on.

llllvvuu commented:
> I think a concrete example would be helpful then, though?

Sure, added

clason commented on runtime/doc/lua.txt:
```diff
@@ -2323,6 +2323,30 @@ vim.tbl_values({t})                                         *vim.tbl_values()*
+
+    Example: >lua
+
+        vim.traverse(
```
```suggestion
        local comments = {}
        vim.traverse(
```

llllvvuu replied:
no worries

MariaSolOs commented on runtime/lua/vim/shared.lua:
```diff
@@ -396,6 +396,37 @@ function vim.tbl_deep_extend(behavior, ...)
+---@param tree any The object to traverse.
+---@param key string|fun(subtree: any): any[] Define the "children" to visit.
+---@param visitor fun(subtree: any): boolean|nil Function to call on each descendant.
+---       Return true to skip the rest of `subtree`.
```
I think it would be helpful to return the `node` if `visitor` returns `true` (and else return `nil`). TypeScript does something like that [here](https://github.com/microsoft/TypeScript/blob/b3770e78527a379d2345412bbab5acc0bafe0090/src/compiler/core.ts#L34).

llllvvuu replied:
Interesting! I thought about `vim.traverse` having a return value; what makes it hard is that you need to apply a reducer since multiple branches may return. Whereas when iterating over a list there is only one branch.

I could make a `tree_reduce` but I think I'd be getting ahead of myself.

I'll think about it some more!

MariaSolOs replied:
Hmmm that's true. There's also the hacky approach of having a special return value for indicating that the traversal must stop completely, like done [here](https://github.com/microsoft/TypeScript/blob/b3770e78527a379d2345412bbab5acc0bafe0090/src/compiler/utilitiesPublic.ts#L717). I don't like this too much though :/

llllvvuu replied:
I think that makes sense for a "find" use-case rather than a "map"/"forEach" use-case; the set of things that gets iterated vs cancelled becomes unpredictable for the user even if they know (leaky abstraction) whether it's DFS, BFS, or iterated DFS

The intention of this PR was to serve the "forEach" use-case and internally we usually use recursion in that way (and so cancelling wants to be of children not of siblings). But I've seen at least one plugin on GitHub Search that used it in a "find"-y way, so who knows, maybe "find_in_tree" or something could be a future utility lol.

MariaSolOs replied:
Sounds reasonable :)

HiPhish commented:
If I may chime in, would it make sense to have an iterator instead or at least in addition? Then we could use a `for` loop to iterate through the nodes of a graph:

```lua
for node in vim.traverse(my_graph, keyfunc) do
    print(vim.inspect(node))
end
```

This seems to be more idiomatic Lua code to me. A visitor function is usually called for its side effects anyway, so the `traverse` function would not compose with other functions anyway.

llllvvuu commented:
Interesting, that sounds like an improved version of https://github.com/neovim/neovim/pull/25154#discussion_r1326804741.

With an iterator you lose the ability to short-circuit a subtree, but you gain the ability to short-circuit an entire traversal.

So I'd make that as two separate functions `vim.traverse` and `vim.iter_tree` (or `vim.iter_graph` with one or two extra lines of impl).

It really depends on Neovim's stance on YAGNI and all that, which is hard to evaluate in an open ecosystem. It sounds like people would use it though, so I'd be happy to make both functions (in the same PR or in separate PR).

HiPhish commented:
> With an iterator you lose the ability to short-circuit a subtree

Do you mean that you want to be able to tell the iterating function "don't descend further down the tree, go back up and resume from there" in the middle of traversal? Could this perhaps be implemented inside the key function?

```lua
local function key_function(node)
	-- short-circuit if some condition is met, pretend that the current node
	-- has no children whatsoever
	if some_predicate(node) then return {} end
	-- Otherwise descend into the children
	return node:get_children()
end
```

llllvvuu commented:
> Could this perhaps be implemented inside the key function?

Based on the recursions that I see in the Neovim codebase, it would probably be duplicated code with the visitor in many cases. So I think it's easiest to split into two APIs and not have one API try to do everything

HiPhish commented on runtime/lua/vim/shared.lua:
```diff
@@ -396,6 +396,37 @@ function vim.tbl_deep_extend(behavior, ...)
+---       Return true to skip the rest of `subtree`.
+function vim.traverse(tree, key, visitor)
+  if not visitor(tree) then
+    local children = type(key) == 'function' and key(tree) or tree[key]
```
In Lua any value is a valid index, including functions, so this check will have an unexpected effect if a tree uses functions as its keys. I know this is a very weird use-case, but I think it is best avoided.

llllvvuu replied:
Are you saying a user might want to pass in a `type(key) == 'function'` but get a `tree[key]`? I think the user can accomplish that with `vim.traverse(tree, function(tree) return tree[key] end, visitor)` instead of `vim.traverse(tree, key, visitor)`.

HiPhish commented:
> Based on the recursions that I see in the Neovim codebase, it would probably be duplicated logic with the visitor in many cases.

I don't quite understand what you mean.

> So I think it's easiest to split into two APIs and not have one API try to do everything

We already have several places where two different APIs do almost but not quite the same, so I really would not like to see this happen again. But at least here they would have radically different use-cases.

llllvvuu commented:
> > Based on the recursions that I see in the Neovim codebase, it would probably be duplicated logic with the visitor in many cases.
> 
> I don't quite understand what you mean.

So many instances of recursion in the Neovim codebase look like:
```lua
function do_recursive_thing(tree)
  do_the_work(tree) -- has side effects, also typically inline
  if work_result then
    for _, subtree in pairs(tree.children) do_recursive_thing() end
  end
end
```

If you had to calculate `work_result` in `key`, then you'd `do_the_work()` twice:

```lua
for subtree in vim.iter_graph(tree,
  function(subtree)
    do_the_work(subtree)
    if work_result then return subtree.children else return {} end
  end)
  do_the_work(subtree)
end
```

You could refactor it to not, but at that point you're better off not using the API. Versus:

```lua
vim.traverse(tree, "children", function(subtree)
  do_the_work(subtree)
  return not work_result
end)
```

llllvvuu commented:
> We already have several places where two different APIs do almost but not quite the same, so I really would not like to see this happen again.

Every stdlib in the world ends up looking like this (especially string/regex/array libs), sometimes the reason is not good, often it is (things that look similar actually being just different enough that trying to abstract over it ends up looking/performing ugly)

clason commented:
Yes, but we can have nice things, even at the cost of deprecating old patterns that have not stood the test of time :)

llllvvuu commented:
Fair enough.

That said, in this particular case, two-in-one seems truly unviable. Iterator vs callback are too fundamentally different as per https://github.com/neovim/neovim/pull/25154#issuecomment-1721895688 and https://github.com/neovim/neovim/pull/25154#issuecomment-1721867254.

HiPhish commented:
OK, I now understand the problem with duplicate logic. The issue is that whether to resume iteration depends on the result of the loop body, but there is no way of passing the result of the current iteration back to the key function. We could move all side effects into the key function and have an empty loop body, but at this point we are back to the function but less readable.

llllvvuu commented:
Current status on this:

* Moved to `vim.func` (although, this is imperative programming)
* leaving as-is as I don't see a way to hybridize traverse with iter/find.
* Will make a separate PR for iter (which also handles find).
* Other improvements to traverse still considered

llllvvuu commented:
Current status on this:

* Leaving as-is as I don't see a way to hybridize traverse with iter/find (open to other improvements though)
* Will make a separate PR for iter (which also handles find)
* Any opinions on location? I put it in `shared.lua` since it feels most similar to `tbl_*` functions; not in `func.lua` since it's imperative programming; and no `tree.lua` or `graph.lua` yet (does it make sense to create that? would just have `vim.graph.traverse()` and `vim.graph.iter()` for now)

justinmk commented:
Thinking about [Iter:join()](https://github.com/neovim/neovim/issues/26409), and how `flatten()` is a special-case of iterable `traverse`...

>  don't see a way to hybridize traverse with iter/find 

Could we return an iterator if `visitor` is nil? @HiPhish @gpanders @lewis6991 

> Any opinions on location?

Seems flexible enough and fundamental enough to live at `vim.traverse()`.
